sonar-kotlin-checks/src/main/java/org/sonarsource/kotlin/checks/AndroidPersistentUniqueIdentifierCheck.kt

```diff
@@ -0,0 +1,108 @@
+private const val HARDWARE_ID_MESSAGE = "Using a hardware identifier puts user privacy at risk. Make sure it is safe here."
+private const val PHONE_NUMBER_MESSAGE = "Using a phone number puts user privacy at risk. Make sure it is safe here."
+private const val ADVERTISING_ID_MESSAGE = "Using Advertising ID puts user privacy at risk. Make sure it is safe here."
+
+private val staticSettingsSecureGetStringFunMatcher = FunMatcher {
+    qualifier = "android.provider.Settings.Secure"
+    name = "getString"
+    withArguments("android.content.ContentResolver", STRING_TYPE)
+}
+private val settingsSecureClassId = ClassId.fromString("android/provider/Settings.Secure")
+private val staticSettingsSecureAndroidIdName = Name.identifier("ANDROID_ID")
+
+private val instanceMatches = mapOf(
+    FunMatcher {
+        qualifier = "android.bluetooth.BluetoothAdapter"
+        name = "getAddress"
+    } to HARDWARE_ID_MESSAGE,
+    FunMatcher {
+        definingSupertype = "android.net.wifi.WifiInfo"
+        name = "getMacAddress"
+    } to HARDWARE_ID_MESSAGE,
+    FunMatcher {
+        definingSupertype = "android.telephony.TelephonyManager"
+        withNames("getSimSerialNumber", "getDeviceId", "getImei", "getMeid")
+    } to HARDWARE_ID_MESSAGE,
+    FunMatcher {
+        definingSupertype = "android.telephony.TelephonyManager"
+        withNames("getLine1Number")
+    } to PHONE_NUMBER_MESSAGE,
+    FunMatcher {
+        definingSupertype = "android.telephony.SubscriptionManager"
+        name = "getPhoneNumber"
+    } to PHONE_NUMBER_MESSAGE,
+    FunMatcher {
+        withDefiningSupertypes(
+            "com.google.android.gms.ads.identifier.AdvertisingIdClient.Info",
+            "androidx.ads.identifier.AdvertisingIdInfo",
+            "com.huawei.hms.ads.identifier.AdvertisingIdClient.Info",
+        )
+        name = "getId"
+    } to ADVERTISING_ID_MESSAGE,
+)
+
+@Rule(key = "S7435")
+class AndroidPersistentUniqueIdentifierCheck : CallAbstractCheck() {
+
+    override val functionsToVisit = instanceMatches.keys
+
+    override fun visitReferenceExpression(expression: KtReferenceExpression, data: KotlinFileContext) = withKaSession {
+        val resolvedCall = expression.resolveToCall()
+        val call = resolvedCall?.singleCallOrNull<KaCallableMemberCall<*, *>>() ?: return@withKaSession
+        if (staticSettingsSecureGetStringFunMatcher.matches(call)) {
+            val nameValueArgument = resolvedCall.singleFunctionCallOrNull()?.argumentMapping?.keys?.toList()?.getOrNull(1)
+                ?: return@withKaSession
+            if (isSettingsSecureAndroidId(nameValueArgument) || nameValueArgument.predictRuntimeStringValue() == "android_id") {
+                data.reportIssue(expression, NON_RESETTABLE_PERSISTENT_ID_MESSAGE)
+            }
+        } else {
+            val matcher = instanceMatches.keys.firstOrNull { it.matches(call) } ?: return@withKaSession
```

Comment:
`keys.firstOrNull` on the dictionary can technically introduce a non-deterministic behavior, if matcher would overlap, so that `firstOrNull` could return different results.
This is however, not the case: matchers are mutually disjoint. An alternative would be to not use a dictionary, but a list of key-value pairs. Makes the code a bit more cumbersome, but I am open to it, if you think it's a better approach.